### ISSUE TITLE
Unable to run GroundTruthReconstruction.py

### ISSUE THREAD

**Issue by Zhentao-Liu** [cicwi/WalnutReconstructionCodes#5]
I cannot run GroundTruthReconstruction.py. It will say that "running power iteration to determine step size", what's wrong.

**Comment by FelixLucka** [comment]
Nothing is wrong: The ground truth reconstruction is done with an iterative reconstruction technique ("accelerated projected gradient descent") that needs a step size to work properly. This step size is determined by a procedure called power iteration. After this is finished, the iterative reconstruction will start. Depending on your GPU, this might take a while.

**Comment by Zhentao-Liu** [comment]
![command](https://github.com/cicwi/WalnutReconstructionCodes/assets/81148025/97ce0523-3fba-4248-a8b5-9184c0a9ca5b)

my command prompt just type the above words, and then just stop.
I think it doesnt finished the work. :<

**Comment by Zhentao-Liu** [comment]
It just stop when running "alg_id = astra.algorithm.create(cfg_agd)"

**Comment by FelixLucka** [comment]
Does FDKReconstruction.py work? If so, how long does it take?

**Comment by Zhentao-Liu** [comment]
Yes, it works well, and takes about 16.703 sec to finish reconstruction (only the reconstruction part).
My test env: python 3.8, win 10, NVIDIA 3060,

**Comment by FelixLucka** [comment]
It seems to get stuck in the for loop in the power iteration (line 37 in NesterovGradient.py). Each iteration could take a few minutes, the whole loop up to an hour. You could use a debugger to step through the code to see where it gets stuck. Alternatively, add some print commands to into the for loop to figure out where it get's stuck

**Comment by Zhentao-Liu** [comment]
Thanks for your guidance. I've found what's wrong.
When I call "alg_id = astra.algorithm.create(cfg_agd)", and get into power_interation in NesterovGradient.py. When I run "b_k1_norm = np.linalg.norm(b_k1)", my programme just stop. Just for 1 second, my GPU reach 100%, and it become normal. (3060 12G). what's your test env.

**Comment by Zhentao-Liu** [comment]
I've solved this issue. np.linalg.norm is too much resource consuming. I batchfy the computation, and then the algorithm can run normally.

def batch_norm(self, input):
    batch_size = 1000000  #adjust to your system capability 
    sum_of_squares = 0
    for i in range(0, len(input), batch_size):
        sum_of_squares += np.sum(input[i:i+batch_size]**2)
    norm = np.sqrt(sum_of_squares)
    return norm
  
hope this would help other people.

**Comment by FelixLucka** [comment]
Thanks!
